### ISSUE TITLE
Improvements enabled by new wat parser

### ISSUE THREAD

**Issue by tlively** [WebAssembly/binaryen#6617]
This is an umbrella tracking issue for further improvements enabled by the new wat parser work. (The parser work itself was tracked in #6208.) 

- #6612
- #6613
- #6614
- #6615
- #6616

**Comment by laptou** [comment]
I think #6407 should be included here too, right?

**Comment by tlively** [comment]
Yes, that was reachable indirectly via #6614, but I've added it explicitly above.
